k8s/operators/src/lib.rs

```diff
@@ -1,2 +1,59 @@
-/// The DiskPool operator
+use crate::error::Error;
+use kube::Client;
+use tracing::{info, warn};
+
+/// The DiskPool operator.
 pub mod diskpool;
+/// The error module for the operator.
+pub mod error;
+/// The MayaStorPool Operator definitions.
+pub mod mayastorpool;
+
+/// Migrate from MayastorPool
+pub async fn migrate_and_clean_msps(k8s: &Client, namespace: &str) -> Result<(), Error> {
+    // Check if the MayastorPool CRD is present, and migrate from it if it is.
+    match mayastorpool::check_crd(k8s).await {
+        // Fetch the MayastorPool CRs.
+        Ok(true) => match mayastorpool::list(k8s, namespace, 100).await {
+            Ok(mut msps) => {
+                for msp in msps.iter() {
+                    let name = msp.metadata.name.as_ref().ok_or(Error::InvalidCRField {
+                        field: "crd.metadata.name".to_string(),
+                    })?;
+                    let node = msp.spec.node();
+                    let disks = msp.spec.disks();
+                    // Create the corresponding DiskPool CRs.
+                    diskpool::create(
+                        k8s,
+                        namespace,
+                        name,
+                        diskpool::DiskPoolSpec::new(node, disks),
+                    )
+                    .await?;
+                }
+                info!("Migration of CRs from MayastorPool to DiskPool complete");
+                // Patch the finalizers and delete the MayastorPool CRs.
+                match mayastorpool::delete(k8s, namespace, &mut msps).await {
+                    Ok(_) => info!("Cleanup of MayastorPool CRs complete"),
+                    Err(error) => {
+                        return Err(Error::Generic {
```

Comment:
Shall we have a non-generic error? Here and below as well?
Also shouldn't we try to migrate and delete all other CR's, otherwise if the first fails we don't try to migrate the other ones?

Reply:
Ah ok.. even so, shouldn't we try to delete all of them rather than fail on the first error?

Reply:
As discussed, we keep continuing on errors and the delete would be along with migration.